dropwizard-oauth/src/main/java/smartthings/dw/oauth/scope/ScopesAllowedDynamicFeature.java

```diff
@@ -12,60 +12,192 @@
 
 		// ScopesAllowed on the method takes precedence over class level annotations
 		ScopesAllowed ra = am.getAnnotation(ScopesAllowed.class);
-		if (ra != null) {
-			configuration.register(new ScopesAllowedRequestFilter(ra.value()));
-			return;
-		}
+		FineGrainedScopesAllowed fgra = am.getAnnotation(FineGrainedScopesAllowed.class);
+
+		if (ra == null) {
+            ra = resourceInfo.getResourceClass().getAnnotation(ScopesAllowed.class);
+        }
 
-		ra = resourceInfo.getResourceClass().getAnnotation(ScopesAllowed.class);
-		if (ra != null) {
-			configuration.register(new ScopesAllowedRequestFilter(ra.value()));
+        if (fgra == null) {
+            fgra = resourceInfo.getResourceClass().getAnnotation(FineGrainedScopesAllowed.class);
+        }
+
+		if ((ra != null) || (fgra != null)) {
+			configuration.register(new ScopesAllowedRequestFilter(ra, fgra));
 		}
 	}
 
 	@Priority(Priorities.AUTHORIZATION) // authorization filter - should go after any authentication filters
 	private static class ScopesAllowedRequestFilter implements ContainerRequestFilter {
 
 		private final String[] scopesAllowed;
+		private final FineGrainedScopeAllowed[] fineGrainedScopesAllowed;
+
+		ScopesAllowedRequestFilter(final ScopesAllowed scopesAllowed,
+                                   final FineGrainedScopesAllowed fineGrainedScopesAllowed) {
+			if ((scopesAllowed != null) && (scopesAllowed.value() != null)) {
+			    this.scopesAllowed = scopesAllowed.value();
+            } else {
+                this.scopesAllowed = new String[]{};
+            }
 
-		ScopesAllowedRequestFilter(final String[] scopesAllowed) {
-			this.scopesAllowed = (scopesAllowed != null) ? scopesAllowed : new String[]{};
+            if ((fineGrainedScopesAllowed != null) && (fineGrainedScopesAllowed.value() != null)) {
+			    this.fineGrainedScopesAllowed = fineGrainedScopesAllowed.value();
+            } else {
+			    this.fineGrainedScopesAllowed = new FineGrainedScopeAllowed[]{};
+            }
+
+            verifyFineGrainedScopes();
 		}
 
 		@Override
 		public void filter(final ContainerRequestContext requestContext) throws IOException {
-			if (scopesAllowed.length > 0 && !isAuthenticated(requestContext)) {
-				throw new ForbiddenException(LocalizationMessages.USER_NOT_AUTHORIZED());
-			}
-
-			Principal principal = requestContext.getSecurityContext().getUserPrincipal();
-			if (principal instanceof ScopeSupport) {
-				for (String role : scopesAllowed) {
-					if (((ScopeSupport) principal).getScopes().contains(role)) {
-						return;
-					}
-				}
-			}
+		    if (isAuthenticated(requestContext)) {
+                Principal principal = requestContext.getSecurityContext().getUserPrincipal();
+                if (principal instanceof ScopeSupport) {
+                    final Collection<String> scopes = Collections.unmodifiableCollection(
+                        ((ScopeSupport) principal).getScopes());
```

Comment:
Is it possible for `principal.getScopes()` to be null?  If so, probably should default to empty list if null.  Otherwise, it's good.